File: app/src/main/java/xyz/jtmiles/beastforgw2/adapters/CharacterInfoAdapter.java
package xyz.jtmiles.beastforgw2.adapters;

import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import org.ocpsoft.prettytime.Duration;
import org.ocpsoft.prettytime.PrettyTime;

import java.util.List;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import xyz.jtmiles.beastforgw2.R;
import xyz.jtmiles.beastforgw2.models.Character;
import xyz.jtmiles.beastforgw2.models.Guild;
import xyz.jtmiles.beastforgw2.services.GuildService;
import xyz.jtmiles.beastforgw2.util.Utils;

/**
 * Created by JT on 5/17/2016.
 */
public class CharacterInfoAdapter extends RecyclerView.Adapter<CharacterInfoAdapter.ViewHolder> {

    private Character mCharacter;

    public CharacterInfoAdapter(Character character) {
        mCharacter = character;
    }

    @Override
    public CharacterInfoAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.character_info_layout, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(final CharacterInfoAdapter.ViewHolder holder, int position) {
        switch(position){
            case 0:
                holder.ivInfoImage.setImageResource(R.drawable.skull);
                holder.tvTitle.setText("Deaths");
                holder.tvContent.setText(String.valueOf(mCharacter.getDeaths()));
                holder.rlLayout.setBackgroundColor(Color.parseColor("#b71c1c"));
                break;
            case 1:
                holder.ivInfoImage.setImageResource(R.drawable.ic_timer_white_48dp_2x);
                holder.tvTitle.setText("Time Played");
                holder.tvContent.setText(String.format(Locale.getDefault(), "%d Hours", mCharacter.getAge() / 60 / 60));
                holder.rlLayout.setBackgroundColor(Color.parseColor("#311b92"));
                break;
            case 2:
                holder.ivInfoImage.setImageResource(R.drawable.ic_people_white_48dp_2x);
                holder.tvTitle.setText("Guild");
                Retrofit retrofit = Utils.getRetrofit(false);
                GuildService guildService = retrofit.create(GuildService.class);
                holder.rlLayout.setBackgroundColor(Color.parseColor("#F9A825"));

                guildService.getGuildDetails(String.format("https://api.guildwars2.com/v1/guild_details.json?guild_id=%s", mCharacter.getGuild())).enqueue(new Callback<Guild>() {
                    @Override
                    public void onResponse(Call<Guild> call, Response<Guild> response) {
                        if (response.isSuccessful()){
                            Guild guild = response.body();
                            holder.tvContent.setText(String.format(Locale.getDefault(), "%s [%s]", guild.getGuildName(), guild.getTag()));

                        }
                    }

                    @Override
                    public void onFailure(Call<Guild> call, Throwable t) {
                        Log.w("CharacterDetailFrag", t.toString());
                    }
                });
                break;
            case 3:
                holder.ivInfoImage.setImageResource(R.drawable.ic_add_white_48dp_2x);
                holder.tvTitle.setText("Created");
                holder.rlLayout.setBackgroundColor(Color.parseColor("#1b5e20"));

                PrettyTime t = new PrettyTime();
                List<Duration> durations = t.calculatePreciseDuration(Utils.getCalendarFromISO(mCharacter.getCreated()).getTime());
                try {
                    durations.remove(durations.size() - 1);
                    durations.remove(durations.size() - 2);
                    durations.remove(durations.size() - 3);
                } catch (ArrayIndexOutOfBoundsException ex) {
                    Log.w("CharacterInfoAdapter", ex);
                }

                holder.tvContent.setText(String.format("%s", t.format(durations)));
                break;
        }
    }

    @Override
    public int getItemCount() {
        return 4;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.ivInfoImage)
        ImageView ivInfoImage;
        @BindView(R.id.tvTitle)
        TextView tvTitle;
        @BindView(R.id.tvContent)
        TextView tvContent;
        @BindView(R.id.rlLayout)
        RelativeLayout rlLayout;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

}

File: app/src/main/java/xyz/jtmiles/beastforgw2/adapters/CharacterInfoAdapter.kt
package xyz.jtmiles.beastforgw2.adapters

import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import org.joda.time.DateTime
import org.joda.time.Period
import org.joda.time.format.PeriodFormatterBuilder
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import xyz.jtmiles.beastforgw2.R
import xyz.jtmiles.beastforgw2.models.Character
import xyz.jtmiles.beastforgw2.models.Guild
import xyz.jtmiles.beastforgw2.services.GuildService
import xyz.jtmiles.beastforgw2.util.Utils
import xyz.jtmiles.beastforgw2.util.bindView
import java.util.*

/**
 * Created by JT on 5/17/2016.
 */
class CharacterInfoAdapter(val mCharacter: Character) : RecyclerView.Adapter<CharacterInfoAdapter.ViewHolder>() {

    var guildService: GuildService? = null

    init {
        guildService = Utils.getRetrofit(false).create(GuildService::class.java)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CharacterInfoAdapter.ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.character_info_layout, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: CharacterInfoAdapter.ViewHolder, position: Int) {
        when (position) {
            0 -> {
                holder.ivInfoImage.setImageResource(R.drawable.skull)
                holder.tvTitle.text = "Deaths"
                holder.tvContent.text = mCharacter.deaths.toString()
                holder.rlLayout.setBackgroundColor(Color.parseColor("#b71c1c"))
            }
            1 -> {
                holder.ivInfoImage.setImageResource(R.drawable.ic_timer_white_48dp_2x)
                holder.tvTitle.text = "Time Played"
                holder.tvContent.text = String.format(Locale.getDefault(), "%d Hours", mCharacter.age!! / 60 / 60)
                holder.rlLayout.setBackgroundColor(Color.parseColor("#311b92"))
            }
            2 -> {
                holder.ivInfoImage.setImageResource(R.drawable.ic_people_white_48dp_2x)
                holder.tvTitle.text = "Guild"
                holder.rlLayout.setBackgroundColor(Color.parseColor("#F9A825"))

                guildService?.getGuildDetails("https://api.guildwars2.com/v1/guild_details.json?guild_id=${mCharacter.guild}")!!.enqueue(object : Callback<Guild> {
                        override fun onResponse(call: Call<Guild>, response: Response<Guild>) {
                        if (response.isSuccessful) {
                            val guild = response.body()
                            holder.tvContent.text = String.format(Locale.getDefault(), "%s [%s]", guild.guildName, guild.tag)

                        }
                    }

                    override fun onFailure(call: Call<Guild>, t: Throwable) {
                        Log.w("CharacterDetailFrag", t.toString())
                    }
                })
            }
            3 -> {
                holder.ivInfoImage.setImageResource(R.drawable.ic_add_white_48dp_2x)
                holder.tvTitle.text = "Created"
                holder.rlLayout.setBackgroundColor(Color.parseColor("#1b5e20"))

                val start = DateTime(mCharacter.created!!)
                val end = DateTime()
                val p = Period(start, end)

                val periodFormatter = PeriodFormatterBuilder()
                        .appendYears().appendSuffix("y").appendSeparator(" ")
                        .appendMonths().appendSuffix("m").appendSeparator(" ")
                        .appendWeeks().appendSuffix("w").appendSeparator(" ")
                        .appendDays().appendSuffix("d").appendSeparator(" ")
                        .toFormatter()

                holder.tvContent.text = p.toString(periodFormatter) + " ago"
            }
        }
    }

    override fun getItemCount(): Int {
        return 4
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val ivInfoImage: ImageView by bindView(R.id.ivInfoImage)
        val tvTitle: TextView by bindView(R.id.tvTitle)
        val tvContent: TextView by bindView(R.id.tvContent)
        val rlLayout: RelativeLayout by bindView(R.id.rlLayout)
    }

}
